### ISSUE TITLE
why to use teacher-forcing in function Validate()?

### ISSUE THREAD

**Issue by 1312657** [sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#6]
Thank you for your implementation of this paper. I think it might be more reasonable that we could not obtain the corresponding caption of an image during validation,  so I could not understand why to use teacher-forcing in function Validate().

**Comment by sgrvinod** [comment]
Thanks for the question! You're right - validation should mimic real inference conditions as much as possible, and should not contain teacher forcing. I used it because it was simpler, and I could compute the validation loss. (Cross Entropy loss is meaningful only when using Teacher Forcing because decoder inputs and lengths can be different). The official Theano repo uses early-stopping on the CE loss (and not the BLEU score, as in the paper), so I assume they did the same.

Teacher Forcing during validation can affect when we early-stop, and also that the validation metric may not actually reflect real performance. The BLEU score, especially, is a metric meant for naturally generated captions. So I've added [`eval.py`](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning/blob/master/eval.py) with which we can compute BLEU scores in a natural inference setting, with or without Beam Search. These results, now listed in the tutorial, indicate that the model did learn quite well.

When I have time, I'll change the validation function to incorporate batched inference without Teacher Forcing (where generation of a caption in a batch stops when it hits `<end>`), and also perhaps batched Beam Search. Early-stopping with BLEU would be completely 'proper' in this case.

**Comment by 1312657** [comment]
@sgrvinod Thank you very much. I am looking forward to your future work.  I sincerely hope you'll take the trouble to discuss it with me if there is any question. Moreover, I have entered the world of PyTorch just for few weeks, so I appreciate it that your tutorial helps me a lot. And I will follow you on GitHub for more excellent work.

**Comment by sgrvinod** [comment]
Thanks! I'm working on the Object Detection and Text Classification tutorials, and they should be up soon. Don't hesitate to post questions or corrections - I'll do my best to respond as soon as possible. It helps me refine the tutorials and my knowledge!

**Comment by malikfahadsarwar** [comment]
can you explain that do we use the teacher forcing in the validation EVEN in order to compute the cross entropy loss not used the autoregressive technique in the validation set?
THANKS
